File: problems/CSES/Dynamic_Programming/coin_combinations_i/coin_combinations_i.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

const int MOD = 1e9 + 7;

int main_v1() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, target;
    cin >> n >> target;

    vector<int> coins(n);
    for (int i = 0; i < n; i++) cin >> coins[i];

    vector<int> dp(target + 1, 0);
    dp[0] = 1;

    for (int curr_sum = 1; curr_sum < target + 1; curr_sum++) {
        for (int coin : coins) {
            if (curr_sum >= coin) {
                dp[curr_sum] += dp[curr_sum - coin];

                if (dp[curr_sum] >= MOD) {
                    dp[curr_sum] -= MOD;
                }
            }
        }
    }

    cout << dp[target] << "\n";

    return 0;
}

int main() {
    int n, x;
    cin >> n >> x;

    vector<int> c(n);
    for (int j = 0; j < n; ++j) {
        cin >> c[j];
    }

    vector<int> ways(x + 1);
    ways[0] = 1;
    for (int i = 1; i <= x; ++i) {
        ways[i] = 0;
        for (int j = 0; j < n; ++j) {
            if (i - c[j] >= 0) {
                ways[i] = (ways[i] + ways[i - c[j]]) % MOD;
            }
        }
    }

    cout << ways[x] << "\n";
}

File: problems/CSES/Dynamic_Programming/coin_combinations_i/coin_combinations_i.py
import sys
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

MOD = 10**9 + 7

# Russian
'''
Решение для этой задачи логически правильное, но изза того что Python не очень быстрый,
легче написать решение для этоц задачи на C++.
'''

# English
'''
The solution to this problem is logically correct, but since Python isn't very fast,
it's easier to write a solution to this problem in C++.
'''

def main():
    n, target = map(int, input().split())
    coins = list(map(int, input().split()))

    dp = [0] * (target + 1)
    dp[0] = 1

    for curr_sum in range(1, target + 1):
        for coin in coins:
            if coin <= curr_sum:
                dp[curr_sum] += dp[curr_sum - coin]

                if dp[curr_sum] > MOD:
                    dp[curr_sum] %= MOD

    print(dp[target])

main()
